### PR TITLE
Ci 358 flag latest and late submissions

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/HMRC.ESFA.Levy.Api.Client/ApprenticeshipLevyApiClient.cs [hunk 1]
```diff
@@ -76,10 +79,25 @@ public ApprenticeshipLevyApiClient(HttpClient client)
...
         }
 
+        public async Task<LevyDeclarations> GetEmployerLevyDeclarationsWithPaymentStatuses(string empRef, DateTime dateRegistered)
+        {
+            var url = $"apprenticeship-levy/epaye/{HttpUtility.UrlEncode(empRef)}/declarations";
+            var parameters = HttpUtility.ParseQueryString(string.Empty);
+
+            if (parameters.AllKeys.Any())
+            {
+                url += "?" + parameters;
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/ApprenticeshipLevyApiClient.cs]: agreed

File: src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace HMRC.ESFA.Levy.Api.Client.Services
+{
+    public class PaymentStatusProcessor : IPaymentStatusProcessor
+    {
+
+        public List<Declaration> ProcessDeclarationPaymentStatuses(List<Declaration> declarations, DateTime dateAdded)
+        {
+            var distinctPeriodsInPayroll = declarations
+                .Select(x => x.PayrollPeriod.Year + x.PayrollPeriod.Month.ToString())
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs]: sounds good, will do.

File: src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                .Distinct();
+
+            foreach (var period in distinctPeriodsInPayroll)
+            {
+                foreach (var declaration in declarations
+                    .Where(x => x.PayrollPeriod.Year + x.PayrollPeriod.Month.ToString() == period)
+                    .OrderByDescending(x => x.SubmissionTime))
+                {
+
+                    var dateOfCutoffForProcessing = DateOfCutoffUtc(declaration.PayrollPeriod, 23);
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs]: yep.

File: src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            
+                    if (dateAdded.ToUniversalTime() >= dateOfCutoffForProcessing) break;
+
+                    var dateOfCutoffForSubmission = DateOfCutoffUtc(declaration.PayrollPeriod, 20);
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs]: yep

File: src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                }
+            }
+
+            return declarations;
+        }
+
+        private static DateTime DateOfCutoffUtc(PayrollPeriod payrollPeriod, int dateOfCutoff)
+        {
+            var monthOfProcessing = payrollPeriod.Month + 4;
+            var yearOfProcessing = 2000 + int.Parse(payrollPeriod.Year.Substring(0, 2));
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs]: yeah, I'll rename the variable, but it is indeed XX-XX format

File: src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            if (monthOfProcessing > 12)
+            {
+                monthOfProcessing = monthOfProcessing - 12;
+                yearOfProcessing = yearOfProcessing + 1;
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs]: ok

File: src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs]: Ah, I misunderstood.  That field is part of an existing object, and I reckon it's already in use elsewhere ( for example the object is used in the return of ApprenticeshipLevyApiClient. GetEmployerLevyDeclarations), so I don't think I should change it without really good reason. I agree the naming is confusing though - it confused me.  I'm going to leave as is, and we can discuss further once Scott C is back - he may be more aware of the ramifications.

File: src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        public List<Declaration> ProcessDeclarationPaymentStatuses(List<Declaration> declarations, DateTime dateAdded)
+        {
+            var declarationsWithPaymentPeriod = declarations.Where(x => x.PayrollPeriod != null).ToList();
+
+            var distinctPeriodsInPayroll = declarationsWithPaymentPeriod
+                .Select(x => $@"{x.PayrollPeriod.Year}{x.PayrollPeriod.Month}")
+                .Distinct();
+
+            foreach (var period in distinctPeriodsInPayroll)
```
Comment (scottcowan) [inline src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs]: this section is complicated, It's a hard to see what it's doing by just looking at it

File: src/HMRC.ESFA.Levy.Api.Client/ApprenticeshipLevyApiClient.cs [hunk 8]
```diff
@@ -15,16 +16,18 @@ namespace HMRC.ESFA.Levy.Api.Client
...
+        private readonly IPaymentStatusProcessor _paymentStatusProcessor;
         private const string DateFormat = "yyyy-MM-dd";
 
         /// <summary>
         /// Default constuctor
         /// </summary>
         /// <param name="client">A configured HttpClient, alternatively use ApprenticeshipLevyApiClient.CreateHttpClient(token, url)</param>
-        public ApprenticeshipLevyApiClient(HttpClient client)
+        /// <param name="paymentStatusProcessor"></param>
+        public ApprenticeshipLevyApiClient(HttpClient client, IPaymentStatusProcessor paymentStatusProcessor)
```
Comment (scottcowan) [inline src/HMRC.ESFA.Levy.Api.Client/ApprenticeshipLevyApiClient.cs]: We should keep the public cstr

```c#
public ApprenticeshipLevyApiClient(HttpClient client)
```

and you can use an internal cstr

```c#
internal ApprenticeshipLevyApiClient(HttpClient client, IPaymentStatusProcessor paymentStatusProcessor)
 : this(client, new PaymentStatusProcessor()){
}
```

File: src/HMRC.ESFA.Levy.Api.UnitTests/ApprenticeshipLevyApiClientTests.cs [hunk 9]
```diff
@@ -16,25 +18,27 @@ namespace HMRC.ESFA.Levy.Api.UnitTests
...
             };
-            var mockHttp = new MockHttpMessageHandler();
+            MockHttpMessageHandler mockHttp = new MockHttpMessageHandler();
             mockHttp.When($"http://localhost/apprenticeship-levy/epaye/{HttpUtility.UrlEncode(expected.EmpRef)}/declarations")
                 .Respond("application/json", JsonConvert.SerializeObject(expected));
 
             var httpClient = mockHttp.ToHttpClient();
             httpClient.BaseAddress = new Uri("http://localhost/");
-            var client = new ApprenticeshipLevyApiClient(httpClient);
+            var client = new ApprenticeshipLevyApiClient(httpClient, null);
```
Comment (sachpatel) [inline src/HMRC.ESFA.Levy.Api.UnitTests/ApprenticeshipLevyApiClientTests.cs]: Mock.Of<IPaymentStatusProcessor>() ??

File: src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                        declaration.LevyDeclarationPaymentStatus = LevyDeclarationPaymentStatus.LatestPayment;
+                        break;
+                    }
+                }
+            }
+
+            return declarations;
+        }
+
+        private static DateTime DateOfCutoffUtc(PayrollPeriod payrollPeriod, int dateOfCutoff)
```
Comment (sachpatel) [inline src/HMRC.ESFA.Levy.Api.Client/Services/PaymentStatusProcessor.cs]: Could this method start with Get in the method name

File: src/HMRC.ESFA.Levy.Api.UnitTests/ApprenticeshipLevyApiClientTests.cs [hunk 11]
```diff
@@ -16,25 +18,27 @@ namespace HMRC.ESFA.Levy.Api.UnitTests
...
             {
                 EmpRef = "000/AA00000",
-                Declarations = new List<Declaration>()
-                {
-                    new Declaration()
-                }
+                Declarations = expectedDeclarations
             };
-            var mockHttp = new MockHttpMessageHandler();
+            MockHttpMessageHandler mockHttp = new MockHttpMessageHandler();
```
Comment (scottcowan) [inline src/HMRC.ESFA.Levy.Api.UnitTests/ApprenticeshipLevyApiClientTests.cs]: does this need to be explicit or can you use var?

File: src/HMRC.ESFA.Levy.Api.UnitTests/ApprenticeshipLevyApiClientTests.cs [hunk 12]
```diff
@@ -60,15 +64,49 @@ public class ApprenticeshipLevyApiClientTests
...
+            var httpClient = mockHttp.ToHttpClient();
+            httpClient.BaseAddress = new Uri("http://localhost/");
+            var client = new ApprenticeshipLevyApiClient(httpClient, mockDeclarationTypeProcessor.Object);
+
+            // Act
+            var declarations = await client.GetEmployerLevyDeclarationsWithPaymentStatuses(expected.EmpRef, new DateTime());
+
+            // Assert
+            mockHttp.VerifyNoOutstandingExpectation();
+            Assert.AreEqual(expected.EmpRef, declarations.EmpRef);
```
Comment (scottcowan) [inline src/HMRC.ESFA.Levy.Api.UnitTests/ApprenticeshipLevyApiClientTests.cs]: you probably want to add

```c#
mockHttp.VerifyAll();
mockDeclarationTypeProcessor.VerifyAll();
```

File: src/HMRC.ESFA.Levy.Api.UnitTests/PaymentStatusesProcessorTests.cs [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        [Test]
+        public void ShouldReturnEmptyListIfPassedEmptyList()
+        {
+            Assert.AreEqual(_emptyDeclarationPostProcessed.Count, 0);
+        }
+
+        [Test]
+        public void ShouldSetLate1PaymentStatusToLate()
+        {
+            var expectedLateEntry = _declarationsPostProcessed.First(x => x.Id == "Late1");
```
Comment (scottcowan) [inline src/HMRC.ESFA.Levy.Api.UnitTests/PaymentStatusesProcessorTests.cs]: There's a potential for a typo by selecting based on a string here. you could use statics for the test data

```c#
                private static Declaration noPayrollPeriod = new Declaration
                {
                     Id = "noPayrollPeriod",
                     SubmissionTime = new DateTime(2017, 8, 19, 00, 00, 00, DateTimeKind.Utc)
                 };
```

File: src/HMRC.ESFA.Levy.Api.UnitTests/PaymentStatusesProcessorTests.cs [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+                SubmissionTime = new DateTime(2017, 8, 01, 00, 00, 00),
+                PayrollPeriod = new PayrollPeriod {Month = 5, Year = "brokenFormat"}
+            }
+        };
+            var dateAccountCreated = new DateTime(2017, 01, 01, 00, 00, 00, DateTimeKind.Utc);
+
+            var processor = new PaymentStatusProcessor();
+            Assert.Throws<FormatException>(() => processor.ProcessDeclarationPaymentStatuses(brokenDeclarationList, dateAccountCreated));
+    }
+}
```
Comment (scottcowan) [inline src/HMRC.ESFA.Levy.Api.UnitTests/PaymentStatusesProcessorTests.cs]: indentation on the
```c#
}
```

File: src/HMRC.ESFA.Levy.Api.Client/ApprenticeshipLevyApiClient.cs (hunk 8 above)
Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/ApprenticeshipLevyApiClient.cs]: but.....

http://chapterbreak.net/wp-content/uploads/2015/12/only-one.gif

Comment (MarkFCain) [reply src/HMRC.ESFA.Levy.Api.Client/ApprenticeshipLevyApiClient.cs]: I'm kind of 50/50 on doing either this, or just concreting up the processor inside the function that uses it, which is equally smelly but a little more honest/direct.  Thoughts?

**Comment by scottcowan** [comment]
If you're putting the statuses back in, could you get the client to take a settings interface in the constructor to manually set the dates. but also allow it to have default dates

File: src/HMRC.ESFA.Levy.Api.UnitTests/PaymentStatusTest.cs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+                Assert.AreEqual(LevyDeclarationPaymentStatus.LatePayment, declarationsPostProcessed.Single(x => x.Id == "lateentry").LevyDeclarationPaymentStatus, $"Check declaration with id lateentry");
+                Assert.AreEqual(LevyDeclarationPaymentStatus.UnprocessedPayment, declarationsPostProcessed.Single(x => x.Id == "latestentry").LevyDeclarationPaymentStatus, $"Check declaration with id latestentry");
+                Assert.AreEqual(LevyDeclarationPaymentStatus.LatestPayment, declarationsPostProcessed.Single(x => x.Id == "anotherlatestentry").LevyDeclarationPaymentStatus, $"Check declaration with id anotherlatestentry");
+                Assert.AreEqual(LevyDeclarationPaymentStatus.UnprocessedPayment, declarationsPostProcessed.Single(x => x.Id == "futurepayrollentry").LevyDeclarationPaymentStatus, $"Check declaration with id futurepayrollentry");
+                Assert.AreEqual(LevyDeclarationPaymentStatus.UnprocessedPayment, declarationsPostProcessed.Single(x => x.Id == "anotherfuturepayrollentry").LevyDeclarationPaymentStatus, $"Check declaration with id anotherfuturepayrollentry");
+            });
+        }
+
+        public short GetPayrollMonth(int month)
```
Comment (mapaux) [inline src/HMRC.ESFA.Levy.Api.UnitTests/PaymentStatusTest.cs]: private?

File: src/HMRC.ESFA.Levy.Api.UnitTests/PaymentStatusTest.cs [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+        {
+            var taxmonth = month - 3;
+            return (short)(taxmonth <= 0 ? month + 9 : taxmonth);
+        }
+        public string GetPayollYear(DateTime date)
```
Comment (mapaux) [inline src/HMRC.ESFA.Levy.Api.UnitTests/PaymentStatusTest.cs]: private?

**Comment by scottcowan** [comment]
you've got failing unit tests on the build

File: src/HMRC.ESFA.Levy.Api.UnitTests/ApprenticeshipLevyApiClientTests.cs [hunk 17]
```diff
@@ -63,12 +67,48 @@ public class ApprenticeshipLevyApiClientTests
...
+            var expectedDeclarations = new List<Declaration>
+            {
+                new Declaration(),
+                new Declaration()
+            };
+            var expected = new LevyDeclarations
+            {
+                EmpRef = "000/AA00000",
+                Declarations = expectedDeclarations
+            };
```
Comment (arvinthseran) [inline src/HMRC.ESFA.Levy.Api.UnitTests/ApprenticeshipLevyApiClientTests.cs]: we should be mocking the variable 'var expected' in line 87 to live like response from hmrc.
`var expected = "{\"empref\": \"222/AA00002\",\"declarations\": [{\"id\": \"22200000212\",\"submissionId\": 0,\"dateCeased\": \"0001-01-01T00:00:00\",\"inactiveFrom\": \"0001-01-01T00:00:00\",\"inactiveTo\": \"0001-01-01T00:00:00\",\"noPaymentForPeriod\": false,\"submissionTime\": \"2018-04-15T12:00:00.000\",\"payrollPeriod\": {\"year\": \"17-18\",\"month\": 12},\"levyDueYTD\": 0.0,\"levyAllowanceForFullYear\": 15000.0}]}";`

File: src/HMRC.ESFA.Levy.Api.Client/Services/CutoffDatesService.cs [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+        }
+
+        public DateTime GetDateTimeForProcessingCutoff(PayrollPeriod payrollPeriod)
+        {
+            return GetDateOfCutoffInUtc(payrollPeriod, DayInMonthForSubmissionProcessing);
+        }
+
+        private static DateTime GetDateOfCutoffInUtc(PayrollPeriod payrollPeriod, int dateOfCutoff)
+        {
+            const int monthModifierToAlignWithCalendarMonths = 4;
```
Comment (mapaux) [inline src/HMRC.ESFA.Levy.Api.Client/Services/CutoffDatesService.cs]: can you declare it outside this method?

File: src/HMRC.ESFA.Levy.Api.Client/SignificantProcessingDates.cs [hunk 19]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace HMRC.ESFA.Levy.Api.Client
+{
+    internal class SignificantProcessingDates
+    {
+        public DateTime DateProcessorInvoked { get; set; }
```
Comment (scottcowan) [inline src/HMRC.ESFA.Levy.Api.Client/SignificantProcessingDates.cs]: the DateProcessorInvoked is a runtime value and the rest are configuration so you should probably break these out

**Pull request merged**